### PR TITLE
Feat/popover

### PR DESCRIPTION
**Pull request by Tarun-Jangra4**

### CODE REVIEW

File: ui-app/client/src/components/Popover/PopoverStyle.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		}
+		.popoverTip.rightTip::before {
+			clip-path: polygon(0 0, 0 100%, 100% 50%);
+		}
+
+		.popoverTip::before {
+			content: ' ';
+			width: 100%;
+			height: 100%;
+			background-color: black;
```
Comment (serialcoder13) [inline ui-app/client/src/components/Popover/PopoverStyle.tsx:47]: Move this background-color to the styleproperties

File: ui-app/client/src/components/Popover/PopoverStyle.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+			top: 0;
+			display: block;
+		}
+		.popoverContainer {
+			// border-radius: 5px;
+			// background-color: lightblue;
+			// width:300px;
+			// height: 300px;
+			// border: 3px solid black;
+		}
```
Comment (serialcoder13) [inline ui-app/client/src/components/Popover/PopoverStyle.tsx:58]: I think these commented lines are required to give the border around the popover container. Remove width and height. Move border-radius, border and background-color (with no default value) to the style properties.

File: ui-app/client/src/components/Popover/popoverProperties.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	popoverTip: {
+		[COMPONENT_STYLE_GROUP_PROPERTIES.background.type]: {
+			...COMPONENT_STYLE_GROUP_PROPERTIES.background,
+			name: 'popoverTipBackground',
+			displayName: 'Popover Container Background',
+			description: 'Popover Container Background',
+			prefix: 'popoverTip',
+			target: ['popoverTip'],
+		},
+	},
```
Comment (serialcoder13) [inline ui-app/client/src/components/Popover/popoverProperties.ts:87]: Also add the border group to the popover container and add it to the component.

File: ui-app/client/src/components/Popover/popoverProperties.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		],
+	},
+	{
+		name: 'showTip',
+		schema: Schema.ofRef(SCHEMA_REF_BOOL_COMP_PROP),
+		displayName: 'Show Popover tip',
+		description: `Show Popover tip.`,
+		defaultValue: true,
+	},
+	COMMON_COMPONENT_PROPERTIES.visibility,
```
Comment (serialcoder13) [inline ui-app/client/src/components/Popover/popoverProperties.ts:60]: add close on leave option.

File: ui-app/client/src/components/Portal.tsx [hunk 5]
```diff
@@ -1,10 +1,11 @@
...
 import ReactDOM from 'react-dom';
-
-type Props = {
+interface Props {
 	children: React.ReactNode;
 	parent?: HTMLElement;
-};
+}
+
 const Portal: React.FC<Props> = ({ children, parent }) => {
```
Comment (serialcoder13) [inline ui-app/client/src/components/Portal.tsx:9]: add an id property.

File: ui-app/client/src/components/Portal.tsx [hunk 6]
```diff
@@ -1,10 +1,11 @@
...
 	const el = React.useMemo(() => document.createElement('div'), []);
```
Comment (serialcoder13) [inline ui-app/client/src/components/Portal.tsx:10]: Don't use the div using the memo function. Create an element only if the parent is null. Add the containing div with the id. Use the memo function to generate a UUID to associate with the newly created div.

Comment (serialcoder13) [reply ui-app/client/src/components/Portal.tsx:10]: On click of this div also the popover should be closed.

File: ui-app/client/src/components/util/getPositions.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
+export default function getPositions(position: String, boxRect: DOMRect, popoverRect: DOMRect) {
```
Comment (serialcoder13) [inline ui-app/client/src/components/util/getPositions.ts:1]: The position data type has to be an enum and create an enum here and use it in the component.

**Review by serialcoder13 — CHANGES_REQUESTED** [review]
Please make the requested changes.

**Review by raja-avinash — APPROVED** [review]
LGTM

**Pull request merged**